vault/plugin_catalog.go

```diff
@@ -129,68 +162,70 @@ func (p *PluginClient) Ping() error {
 	return nil
 }
 
-func (c *Core) setupPluginCatalog(ctx context.Context) error {
-	c.pluginCatalog = &PluginCatalog{
-		builtinRegistry: c.builtinRegistry,
-		catalogView:     NewBarrierView(c.barrier, pluginCatalogPath),
-		directory:       c.pluginDirectory,
-		logger:          c.logger,
+func (c *PluginCatalog) removePluginClient(name, id string) error {
+	var err error
+	mpc, ok := c.multiplexedClients[name]
+	if !ok {
+		return fmt.Errorf("plugin client not found")
 	}
 
-	// Run upgrade if untyped plugins exist
-	err := c.pluginCatalog.UpgradePlugins(ctx, c.logger)
+	pluginClient := mpc.connections[id]
+	multiplexingSupport, err := pluginClient.MultiplexingSupport()
 	if err != nil {
-		c.logger.Error("error while upgrading plugin storage", "error", err)
+		return err
 	}
 
-	if c.logger.IsInfo() {
-		c.logger.Info("successfully setup plugin catalog", "plugin-directory", c.pluginDirectory)
-	}
+	delete(mpc.connections, id)
+	if !multiplexingSupport {
+		pluginClient.client.Kill()
+		err = pluginClient.protocol.Close()
 
-	return nil
+		if len(mpc.connections) == 0 {
+			delete(c.multiplexedClients, name)
+		}
+	} else if len(mpc.connections) == 0 {
```

Comment:
On the multiplexed case, do we want to always call pluginClient.clientConn.Close() to close the corresponding connection? I noticed that the gRPC server doesn't clean up the instances until `gRPCServer.Close` is called so we might be accumulating instances over time if we never reach to the point where the connections map on the core map becomes empty.

Reply:
I tried adding `pluginClient.clientConn.(*pluginClientConn).ClientConn.Close()` to be called each time but I am getting errors.

I need to take a closer look at the server side instances to see exactly what is going on during the plugin lifecycles.

Reply:
The database backend will handle calling Close on the `databaseVersionWrapper` that holds the `v5 v5.Database` client. That will call then call the gRPCServer.Close to cleanup the instances. 

See [(*dbPluginInstance).Close()](https://github.com/hashicorp/vault/blob/84fd330eef22b27261cd8414493a22e0e698cb9e/builtin/logical/database/backend.go#L39) and [(*databaseBackend).clearConnection()](https://github.com/hashicorp/vault/blob/84fd330eef22b27261cd8414493a22e0e698cb9e/builtin/logical/database/backend.go#L309)

I have verified that we do not accumulate instances on the grpc server intances map.